### PR TITLE
Implement Search Result and New Tab Page ads attribution if ads are enabled or disabled

### PR DESCRIPTION
**Pull request by tmancey**
Resolves https://github.com/brave/brave-browser/issues/29888

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run lint`, `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [x] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

## Test Plan:

If `BraveAds` feature is enabled and `should_always_run_service` is set to `true`:
- Confirm purchase intent, anti-targeting, subdivision-targeting, epsilon greedy bandit, issuers and catalog resources are only loaded if Brave Private Ads and/or Brave News are enabled.
- Confirm purchase intent, anti-targeting, subdivision-targeting, epsilon greedy bandit. issuers and catalog resources are unloaded if Brave Private Ads and Brave News are disabled.
- Confirm text classification and text embedding resources are only loaded if Brave Private Ads are enabled.
- Confirm text classification and text embedding resources are unloaded if Brave Private Ads is disabled.
- Confirm conversion resource is always loaded.
- Confirm ads are served after relaunching the browser.
- Confirm updated resource components are loaded.
- Confirm ads are served after gaining the users attention on desktop.
- Confirm ads are served periodically on mobile.
- Confirm conversions are working as expected.
- Confirm user activity is working as expected.
- Confirm ad events are working as expected, i.e. viewed, clicked etc. are sent to the server.
- Confirm users are only rewarded for ad impressions if they have enabled Brave Private Ads.
- Confirm search result ad events are triggered, such as viewed, clicked, conversion etc.
- Confirm history is  onlyadded to brave://rewards if Brave Private Ares are enabled

If `BraveAds` feature is enabled and `should_always_run_service` is set to `false`:
- Confirm purchase intent, anti-targeting, subdivision-targeting, epsilon greedy bandit, conversions, issuers and catalog resources are only loaded if Brave Private Ads and/or Brave News are enabled.
- Confirm text classification and text embedding resources are only loaded if Brave Private Ads are enabled.
- Confirm text classification and text embedding resources are unloaded if Brave Private Ads is disabled and Brave News is enabled.
- Confirm ads are served after relaunching the browser.
- Confirm updated resource components are loaded.
- Confirm ads are served after gaining the users attention on desktop.
- Confirm ads are served periodically on mobile.
- Confirm conversions are working as expected.
- Confirm user activity is working as expected.
- Confirm ad events are working as expected, i.e. viewed, clicked etc. are sent to the server.
- Confirm users are only rewarded for ad impressions if they have enabled Brave Private Ads.
- Confirm search result ad events are triggered, such as viewed, clicked, conversion etc.
- Confirm history is only added to brave://rewards if Brave Private Ares are enabled

### CODE REVIEW

File: components/brave_ads/core/internal/conversions/conversions.cc [hunk 1]
```diff
@@ -47,19 +48,34 @@ bool HasObservationWindowForAdEventExpired(
...
+    }
+
+    case AdType::kSearchResultAd: {
+      // Always convert.
+      return true;
+    }
+
+    case AdType::kUndefined: {
+      NOTREACHED_NORETURN();
+    }
```
[the remark this answers is not in the document]
Comment (fmarier) [reply components/brave_ads/core/internal/conversions/conversions.cc:74]: False positive, that macro triggers an immediate crash.

@thypon Does this rule need to be amended with something like:
```
      - pattern-not: |
          switch ($VAR) { case $VAL1: NOT_REACHED_NORETURN() ; case $VAL2: ... }
```

File: components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.cc [hunk 2]
```diff
@@ -98,7 +105,16 @@ void SearchResultAdHandler::MaybeTriggerSearchResultAdClickedEvent(
   }
 
   ads_service_->TriggerSearchResultAdEvent(
-      search_result_ad->Clone(), mojom::SearchResultAdEventType::kClicked);
+      search_result_ad->Clone(), mojom::SearchResultAdEventType::kClicked,
+      /*intentional*/ base::DoNothing());
+}
+
+bool SearchResultAdHandler::IsEnabled() {
+  if (!ads_service_) {
```
Comment (aseren) [inline components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.cc:113]: let's store ads service as a raw_ref<>

File: components/brave_ads/core/internal/account/issuers/issuers.cc [hunk 3]
```diff
@@ -92,13 +104,24 @@ void Issuers::FetchCallback(const mojom::UrlResponseInfo& url_response) {
...
+}
+
 void Issuers::SuccessfullyFetchedIssuers(const IssuersInfo& issuers) {
   StopRetrying();
 
   if (delegate_) {
     delegate_->OnDidFetchIssuers(issuers);
   }
 
+  is_initialized_ = true;
```
Comment (aseren) [inline components/brave_ads/core/internal/account/issuers/issuers.cc:123]: maybe rename to somethink like was_successfully_fetched_

File: components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.cc (hunk 2 above)
Comment (tmancey) [reply components/brave_ads/content/browser/search_result_ad/search_result_ad_handler.cc:113]: Fixed

File: components/brave_ads/core/internal/account/issuers/issuers.cc (hunk 3 above)
Comment (tmancey) [reply components/brave_ads/core/internal/account/issuers/issuers.cc:123]: Changed, thanks

File: components/ntp_background_images/browser/view_counter_service.h [hunk 4]
```diff
@@ -175,6 +175,8 @@ class ViewCounterService : public KeyedService,
   std::unique_ptr<WeeklyStorage> branded_new_tab_count_state_;
 
   std::unique_ptr<NTPP3AHelper> ntp_p3a_helper_;
+
+  base::WeakPtrFactory<ViewCounterService> weak_factory_{this};
```
Comment (aseren) [inline components/ntp_background_images/browser/view_counter_service.h:179]: seems like not needed

**Review by aseren — APPROVED** [review]

File: components/ntp_background_images/browser/view_counter_service.h (hunk 4 above)
Comment (tmancey) [reply components/ntp_background_images/browser/view_counter_service.h]: Fixed :-)

**Review by Brandon-T — APPROVED** [review]

**Review by szilardszaloki — APPROVED** [review]
Rewards changes look good! 👍🏻

File: components/brave_ads/browser/ads_service_impl.cc [hunk 5]
```diff
@@ -309,15 +315,32 @@ AdsServiceImpl::AdsServiceImpl(
...
+
   rewards_service_->AddObserver(this);
+
+  if (kShouldAlwaysRunService.Get()) {
+    RegisterResourceComponentsForDefaultLocale();
+  }
 }
 
 AdsServiceImpl::~AdsServiceImpl() {
+  g_brave_browser_process->resource_component()->RemoveObserver(this);
```
Comment (bridiver) [inline components/brave_ads/browser/ads_service_impl.cc:328]: this is a layering violation, you need to pass `g_brave_browser_process->resource_component()` to the constructor

**Pull request merged**